workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Utils/CodeWriter.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Text;
+
+namespace Improbable.Gdk.CodeGenerator.Utils
+{
+    public class CodeWriter
+    {
+        private int scopeCounter;
+        private bool hasBuiltAlready;
+        private StringBuilder builder;
+
+        private const int TabLength = 4;
+        
+        public CodeWriter()
+        {
+            builder = new StringBuilder();
+        }
+
+        public CodeScope Scope()
+        {
+            return new CodeScope(this);
+        }
+
+        public void WriteLine(string line)
+        {
+            Indent();
+            builder.AppendLine(line);
+        }
+
+        public string Build()
+        {
+            if (scopeCounter != 0)
+            {
+                throw new InvalidOperationException("Cannot build when scope is not at the root scope.");
+            }
+
+            if (hasBuiltAlready)
+            {
+                throw new InvalidOperationException("Cannot call Finish() more than once on the same CodeWriter instance");
+            }
+
+            hasBuiltAlready = true;
+            return builder.ToString();
+        }
+        
+        private void Indent()
+        {
+            builder.Append(new String(' ', TabLength * scopeCounter));
+        }
+        
+        public struct CodeScope : IDisposable
+        {
+            private CodeWriter writer;
+        
+            public CodeScope(CodeWriter writer)
+            {
+                this.writer = writer;
+                
+                writer.WriteLine("{");
+                writer.scopeCounter++;
+            }
+
+            public void Dispose()
```

Comment:
Check if writer is null, and set writer to null after dispose to avoid double dispose